File: chapters/02-design-patterns/java/StrategyNotificationChannels.java
// Scenario: order-confirmation notifications (email, SMS, push)
// Demonstrates: Strategy — ChannelSender per channel without editing dispatcher
// Trade-off: registry vs switch; ensure selection is observable (metrics/logs)

import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

interface ChannelSender {
    String channel();
    void send(String userId, String body);
}

final class EmailSender implements ChannelSender {
    public String channel() { return "email"; }
    public void send(String userId, String body) { /* SES/SMTP */ }
}

final class NotificationDispatcher {
    private final Map<String, ChannelSender> senders;
    NotificationDispatcher(java.util.List<ChannelSender> list) {
        this.senders = list.stream().collect(Collectors.toMap(ChannelSender::channel, Function.identity()));
    }
    void notify(String channel, String userId, String body) {
        ChannelSender s = senders.get(channel);
        if (s == null) throw new IllegalArgumentException("unknown channel: " + channel);
        s.send(userId, body);
    }
}

File: docs/chapters/02-design-patterns/go/strategy_notification_channels.go
// Scenario: order-confirmation notifications (email, SMS, push)
// Demonstrates: Strategy — ChannelSender per channel without editing dispatcher
// Trade-off: map registry vs switch; Go favors small interfaces + explicit errors

package main

import "fmt"

type ChannelSender interface {
	Channel() string
	Send(userID, body string) error
}

type emailSender struct{}

func (emailSender) Channel() string { return "email" }
func (emailSender) Send(_, _ string) error { return nil }

type NotificationDispatcher struct {
	senders map[string]ChannelSender
}

func NewNotificationDispatcher(list []ChannelSender) *NotificationDispatcher {
	m := make(map[string]ChannelSender, len(list))
	for _, s := range list {
		m[s.Channel()] = s
	}
	return &NotificationDispatcher{senders: m}
}

func (d *NotificationDispatcher) Notify(channel, userID, body string) error {
	s, ok := d.senders[channel]
	if !ok {
		return fmt.Errorf("unknown channel: %s", channel)
	}
	return s.Send(userID, body)
}
